### ISSUE TITLE
Address semantic issues related to OrdinalCollection and OrdinalMember

### ISSUE THREAD

**Issue by rjyounes** [semanticarts/gist#112]
OrdinalMember is not so much a _type_ of thing as the relationship between a thing and an OrdinalCollection. Something can be a member of more than one OrdinalCollection just as it can be a member of more than one Collection in general. For example, 3 is a member of the ordinal collection of integers and of prime numbers.

Granted, the semantics of OrdinalCollection will not be as tightly constrained as they are now without the OrdinalMember class, but they will be more accurate.

This also requires that precedence relationships not be stated directly between the members, but through ordering proxies -- e.g., the [OAI-ORE ontology](http://www.openarchives.org/ore/1.0/vocabulary). That is, in the collection of integers 3 directly precedes 4, but in the collection of prime numbers it directly precedes 5. The relationships don't make sense unless stated relative to a specific collection.

On the other hand, the intent of the OrdinalCollection and OrdinalMember classes may be to limit them to sequences like the months of the year or days of the week, where members are part of _only one_ (well-established) collection. In that case, we must say that the sequences of integers and primes are OrderedCollections but not OrdinalCollections. Then the restriction on OrdinalMember that it can be part of only one OrdinalCollection should be formalized in the OWL definition since it's crucial to the semantics.

**Comment by uscholdm** [comment]
*   OrdinalMember is not a type of thing;
Yes, OrdinalMember is more a role that something plays. It is a class so it can be used in restrictions, domains and ranges. Not ideal. There is probably a better way.
There is nothing interesting


  *   there's a relationship between a thing and an OrdinalCollection
There is nothing interesting about the relationship between and OrdinalMember and an OrdinalCollection:  its gist:memberOf.
An OrdinalCollection is meant to be used as a set of possible values for a measure that are ordered.


  *   This also requires that precedence relationships not be stated directly between the members, but through ordering proxies. That is, in the collection of integers 3 directly precedes 4, but in the collection of prime numbers it directly precedes 5. The relationships don't make sense unless stated relative to a specific collection.
  *   And we should state this restriction on OrdinalMember that it can be part of only one OrdinalCollection.
Excellent points.  Do you have any specific suggestions to improve this?

**Comment by johnwcowan** [comment]
I think the way forward is to have a class Ordering that describes the order of elements.  Then an OrderedCollection is a pair <Collection, Ordering>.

**Comment by marksem** [comment]
Related: https://github.com/semanticarts/gist/issues/142

**Comment by rjyounes** [comment]
@johnwcowan I would instead propose a proxy object that sits between a collection and a member which specifies the order. This is the [OAI-ORE](http://www.openarchives.org/ore/1.0/vocabulary) approach. It's easier to query what position a member occupies in the list. Further, if you have an infinite ordered collection, such as the set of positive integers, you cannot actually provide an Ordering, while you can state that the integer 245678 occupies position 245678.

Sample RDF (this example is a little awkward because 5 is a literal and we probably don't want to autobox it in an object, so we cannot make an inverse assertion from 5 to the proxy.

```
:positiveIntegers a gist:OrdinalCollection .
:proxyFor5InPositiveIntegers gist:memberOf :positiveIntegers ;
    :sequence "5"^^xsd:int ;
    gist:decimalValue "5"^^xsd:decimal .

:primeNumbers a gist:OrdinalCollection .
:proxyFor5InPrimeNumbers gist:memberOf :primeNumbers ;
     :sequence "4"^^xsd:int ;
     gist:decimalValue "5"^^xsd:decimal .
```

**Comment by uscholdm** [comment]
Currently, hasOrderedMember is inverse functional. This prevents having one item be in more than one OrderedCollection. A way to get around this is as follows. Lets say:
-  we have 10 objects: Obj1 through Obj10 and 3 lists: OrdList1, Ordlist2 and OrdList3.   
- Obj1 is first in OrdList2, and is in third place in OrdList1 and OrdList3. 

One way to model this is to create a class called say `ItemInSequence`.  To model the above situation we would have two instances of this class -- one for Obj1 in first place and one for it in third place.  In Turtle (and Manchester)
```
OrdList1 a gist:OrderedCollection .
OrdList2 a gist:OrderedCollection .
OrdList3 a gist:OrderedCollection .

Obj1_1 a :ItemInSequence ; gist:sequence 1, gist:memberOf OrdList2.
Obj1_3 a :ItemInSequence ; gist:sequence 3, gist:memberOf OrdList1 , OrdList3.
```

ItemInSequence is used instead of the bogus OrderedMember, which is now gone from gist.

For OrderedCollections formed in this way, we could have a restriction requiring every member to be instance of ItemInSequence. 
E.g. 
`OrderedCollection rdfs:subClassOf (hasMember only ItemInSequence)`

Or maybe better, in SHACL because another way to create an OrderedCollection is to use the precedes/precededBy property. In that case there would be ordinary objects as members. Their sequence could be inferred. 

There is a downside to this: if you want to reorder things in the list, its a big pain.
@MarkWallace encountered this a while back. Mark: what was your experience?

**Comment by rjyounes** [comment]
@uscholdm I believe you are missing the relationship between Obj1 and the item in sequence Obj1_1. If that's added in, this is the same as the proxy approach I've outlined above, unless there's some distinction I've missed.

**Comment by johnwcowan** [comment]
Yes, that confused me too.  You also could have two subclasses of
SequenceItem (better than ItemInSequence, I think), one being
LinkedSequenceItem where the SequenceItems are connected by
precedes/precededBy (or just precedes/follows), and the other being
NumberedSequenceItem, which uses sequence numbers.  These correspond to
linked lists and vectors: one is good for local insertion and deletion, the
other is good for access to the nth element.

**Comment by rjyounes** [comment]
I like the idea of using precedes/precededBy for insertion and deletion. Not in favor of dual methods for representing the same semantics, and especially not on the same collection since duplication is a recipe for getting things out of sync.

**Comment by marksem** [comment]
May want to have a look at OLO (ordered list ontology).  It has a "slot" concept that allows items to be in more than 1 ordered collection. 
http://smiy.sourceforge.net/olo/spec/orderedlistontology.html

**Comment by daliadahleh** [comment]
> If that's added in, this is the same as the proxy approach I've outlined above, unless there's some distinction I've missed.

I believe the difference is that Michael is using an instance per position, rather than an instance per collection. I.e. the instance Obj1_3 can be used in multiple collections as long as it has position 3 in all of those collections.

I prefer an instance per collection, even though it means a slightly larger number of triples, mostly because I prefer to use directlyPrecedes rather than sequence. If you prefer to use sequence, I don't see why an instance per position wouldn't work.

**Comment by rjyounes** [comment]
While this may be _technically_ true, my objection is that this is a semantically unpalatable type of object: an instance which happens to be in 3rd position in any collection. In other words, the notion of "March when it is in 3rd position" in the collection Months vs "March when it is in 2nd position" as in the collection of Months with 31 Days makes no sense to me. March is March. And you would need a predicate that would relate the two Marches, so that they could be as unconnected as March and November; thus expanding the number of triples. In any case, the number of triples is of less significance than correctly capturing the semantics, and the proxy solution is preferable in this regard.

**Comment by johnwcowan** [comment]
Oh, definitely not in the same collection!  There would need to be two
classes of OrderedCollection, too.

But the difference between a numbered list and an unnumbered, but still
ordered, list is substantial, and it's clear that people have different
preferences from a modeling perspective, as well as from the perspective of
making changes.  I think there is a case for doing both.

**Comment by uscholdm** [comment]
> I believe the difference is that Michael is using an instance per position, rather than an instance per collection. I.e. the instance Obj1_3 can be used in multiple collections as long as it has position 3 in all of those collections.

Exactly.  It's just a technique for getting a job done. Not very nice, these objects are artificial.  Am happy to entertain a concrete proposal illustrated with triples to create an Ordered and OrdinalCollections where one item can be a member of more than one collection.  The property hasOrdinalMember is currently inverse functional. There was a good reason for this, but I do not recall what it was.

**Comment by rjyounes** [comment]
@uscholdm See the [OAI-ORE ontology](http://www.openarchives.org/ore/1.0/vocabulary). @marksem also mentions the [Ordered List Ontology](http://smiy.sourceforge.net/olo/spec/orderedlistontology.html).

A turtle fragment looks like this (using invented terms, no particular ontology):

```
:oc1 a :OrderedCollection ;
    :hasMember :oc1_proxy1 , :oc1_proxy2 ; oc1_proxy3 .
:oc1_proxy1 :directlyPrecedes :oc1_proxy2 ;
    :proxyFor :o1 .
:oc1_proxy2 :directlyPrecedes :oc1_proxy3 ;
    :proxyFor :o2 .
:oc1_proxy3 :proxyFor :o3 .

:oc2 a :OrderedCollection ;
    :hasMember :oc2_proxy1 , :oc2_proxy2 ; oc2_proxy3 .
:oc2_proxy1 :directlyPrecedes :oc2_proxy2 ;
    :proxyFor :o3 .
:oc2_proxy2 :directlyPrecedes :oc2_proxy3 ;
    :proxyFor :o1 .
:oc2_proxy3 :proxyFor :o1 .
```

**Comment by rjyounes** [comment]
At any rate, we have not determined whether to use external ontologies in our modeling, so that is a different but relevant topic (there might be an issue for this somewhere but I can't find it). So perhaps the best path is to define our own terms based on the ttl sample shown above.

**Comment by uscholdm** [comment]
> At any rate, we have not determined whether to use external ontologies in our modeling

Rarely would we import an external ontology in toto. There is almost always some clutter or specific issues (like too narrow domains/ranges).  More likely we would adapt an idea and fold it into gist, giving credit to a idea source as needed.  The proxy approach above is workable, as is the olo ontology which is similar to my proposal. Some blend of these is likely to suit us best.

**Comment by rjyounes** [comment]
Actually, the OLO ontology is very close to OAI-ORE and what I've illustrated above. The problem with OLO, as @daliadahleh has pointed out, is that the cardinality requirements do not match our use cases. I believe the proxy approach is the best solution based on the semantic arguments I've made above.

**Comment by uscholdm** [comment]
> @marksem also raises the issue of weak vs strong ordering:

I used to have a property analogous to greater or equal to for this. It mean precedes or is in the same order as. That allows for partial or total ordering.  Dave removed it. I documented the fact that this was removed in a gist council draft deck that Dave edited and gave. He removed the change about precedes and sameOrderAs.  We could bring it back it there was demand.

**Comment by rjyounes** [comment]
See also #114, closed as a duplicate, but may have some interesting discussion.

**Comment by rjyounes** [comment]
Possibly useful reference: https://www.slideshare.net/albertmeronyo/modelling-and-querying-lists-in-rdf-a-pragmatic-study

**Comment by rjyounes** [comment]
Meika's report:
- Consensus forming around proxy solution
- OLO and OAI-ORE seem about the same under the hood; OLO may be simpler and easier to implement in gist

Meika and Mark will:
- create Turtle sample for OLO as above for OAI-ORE
- propose terms

**Comment by rjyounes** [comment]
Reassign to Michael and Boris since needed for Platts.

**Comment by uscholdm** [comment]
FROM @mkumba 

> It looks like gist:OrdinalMember in addition to a restriction is also a Category.  I don’t think that was intentional, was it?

**Comment by rjyounes** [comment]
@uschold Before you added this note, I created a new issue #426. The current issue involves a large chunk of work; I was thinking we could correct this small error before the larger piece gets done. What do you think?

**Comment by rjyounes** [comment]
Dalia will bring back a proposal for further group review.

**Comment by rjyounes** [comment]
Going back a ways in the comment thread, Michael's proposal uses "an instance per position, rather than an instance per collection. I.e. the instance Obj1_3 can be used in multiple collections as long as it has position 3 in all of those collections."

An advantage of the reverse strategy - where the proxy individual relates to a specific object and its position in a specific sequence, rather than any object in a specific position in any sequence - is that you can attach properties to the proxy - e.g., date the item was added to the sequence.

**Comment by rjyounes** [comment]
The agreed-upon implementation is a major change, so deferred till 10.0.0

**Comment by daliadahleh** [comment]
Decision

**Remove:**
Classes gist:OrdinalCollection and gist:OrdinalMember (in favor of gist:OrderedCollection and gist:OrderedMember)

**Add:**
Class gist:OrderedMember - acts directly as the proxy (union of restrictions gist:directlyPrecedes and gist:sequence)
Property gist:isProxyFor



Example triples:

**Using gist:sequence**
```
:_collection1    
    a gist:OrderedCollection ;
    gist:hasOrderedMember 	:_proxy_itemA_collection1 , :_proxy_itemB_collection1 .
	
:_proxy_itemA_collection1	 
    a	gist:OrderedMember ;
    gist:isProxyFor		:_itemA ;
    gist:sequence 		1 .

:_proxy_itemB_collection1	
    a 	gist:OrderedMember ;
    gist:isProxyFor		:_itemB ;
    gist:sequence 		2 .
```

**Using gist:directlyPrecedes**
```
:_collection2    
    a gist:OrderedCollection ;
    gist:hasOrderedMember 	:_proxy_itemA_collection2 , :_proxy_itemB_collection2 .
	
:_proxy_itemA_collection2	 
    a	gist:OrderedMember ;
    gist:isProxyFor		:_itemA ;
    gist:directlyPrecedes	:_proxy_itemB_collection2 .

:_proxy_itemB_collection2	
    a 	gist:OrderedMember ;
    gist:isProxyFor		:_itemB .

**Comment by philblackwood** [comment]
Mathematically, a sequencing is a function from a set to the positive integers.

We want to say: "_thing has position n within collection X", which is an n-ary relationship, so it uses a skip-over kind of relationship between a _thing and the collection, similar to an aspect of a magnitude actually referring to the original thing.

The names below might not be quite right, but one pattern to express the relationship is:

_collectionX   hasOrdinalMember  _position, 

                  links up with  _position hasMember _thing
                                         _position ordinal         _integer (n)

CollectionX fans out into a set of _positions, where each _position is an ordered pair relating a member to its ordinal number within the collection. 

This puts no restriction on how many ordered collections a thing can belong to,.
It creates a limit of one ordering per collection (just create another collection with the same elements). 
Works fine for infinite sequences with a clear starting point. 
Can also work with infinite sequences extending in two directions (positive and negative) if you stretch the definition of "ordinal" to allow negative integers.

Apologies in advance if I'm repeating previous comments.

This is kind of off the top of my head, so let me know if I need to go back to the drawing board.

**Comment by rjyounes** [comment]
@daliadahleh "Class gist:OrderedMember - acts directly as the proxy (union of restrictions gist:directlyPrecedes and gist:sequence)" I believe should read:

Class gist:OrderedMember - subclass of (union of restrictions gist:directlyPrecedes and gist:sequence and gist:isDirectlyPrecededBy)

If this is correct, please add to above summary.

**Comment by rjyounes** [comment]
What about the (edge) case where all members have the same rank - e.g., a record of race results where everyone has the same time? (This is not a fantasy; it can sort of happen in bicycle races.) We still want to consider it an `OrderedCollection` because we don't know in advance that this will be the result, and it could have turned out otherwise, but neither `directlyPrecedes` nor `directlyPrecededBy` apply to any of the members. Do we say that in that case a `sequence` would be stated, which is then still covered by the restriction on the union (`directlyPrecedes` U `directlyPrecededBy` U `sequence`)?

**Comment by mkumba** [comment]
Yeah I think we just give two of the ordered members sequence 1

On Apr 9, 2021, at 6:54 AM, Rebecca Younes ***@***.******@***.***>> wrote:



What about the (edge) case where all members have the same rank - e.g., a record of race results where everyone has the same time? (This is not a fantasy; it can sort of happen in bicycle races.) We still want to consider it an OrderedCollection because we don't know in advance that this will be the result, and it could have turned out otherwise, but neither directlyPrecedes nor directlyPrecededBy apply to any of the members. Do we say that in that case a sequence would be stated, which is then still covered by the restriction on the union (directlyPrecedes U directlyPrecededBy U sequence)?

**Comment by philblackwood** [comment]
It looks like mathematical foundation validates solution ... and it looks like the proxy pattern has other uses.

See diagrams in attached PowerPoint (5 minute read).
[mathematical relation.pptx](https://github.com/semanticarts/gist/files/6287781/mathematical.relation.pptx)

**Comment by uscholdm** [comment]
> but neither directlyPrecedes nor directlyPrecededBy apply to any of the members.

This is not a problem if we treat `directlyPrecedes` and `directlyPrecededBy` to include being the `sameOrderAs`. This is analogous to less than or equal to instead of strictly less than. I had this in gist at one point, but it was removed after years of non-use.

**Comment by philblackwood** [comment]
gist issue #112 in pictures. Touches on many of the comments posted about the issue.

Might help clarify "when to use what" for ordered collections and context-specific data.

Not intended to propose anything, just state some useful facts, and a few opinions.

Not intended to be a quick or easy read. Hats off to you if it is.

[gist issue 112 in pictures.pptx](https://github.com/semanticarts/gist/files/6300057/gist.issue.112.in.pictures.pptx)

**Comment by philblackwood** [comment]
Did somebody say "vector"?   (seeking input, and offering clarification)

Back to set theory, there are two very useful but distinct patterns relating individuals to integers.

1. start with a collection, assign each an integer.  Four people run for election:
            
               Bob.     1
               Amy.    2
               Joe       2
               Deb      3

2. start with a list of integers. relate each to an individual

               1     Bob
               2.    Amy
               3.    Bob
               4.    Deb

1 is a function from a collection to the integers, and 2 is a function from some integers to a collection.

The second pattern can be thought of as a vector, or a playlist.   I think rdf calls this a Seq. (rdf:Seq class, with rdf:_1, rdf:_2, etc.)

I'm looking for a good representation for vectors, because every derived coherent unit of measure can be represented as a vector of exponents (once you put the base units in a specific order).  The dimension corresponding to the derived unit can be represented by the same vector (once you put the base dimensions in the same sequence as their corresponding base units).

Any suggestions for working with vectors in OWL?

Could somebody please invent playlists for Powerpoint slides?  :-)

**Comment by sa-bpelakh** [comment]
So, to throw another spanner in the works - how do support an ordered list of literals? This just occurred to me upon examining the lovely `msrpc:focusDescription_X` set of DatatypeProperties at MorganStanley. Does `proxyFor` require a range of `rdf:Resource` (much like `rdf:first`) so that a list of literals can be handled?

**Comment by marksem** [comment]
(To clarify, an rdfs:Resource can also be a literal. Ref [1]
Also, there is no rdf:Resource defined in the RDF nor RDFS spec that I can find. There is rdfs:Resource) 

So I would imagine you can have proxyFor point at a literal just as easy as a IRI or blank node. 

[1] https://www.w3.org/TR/rdf11-concepts/#resources-and-statements
[2] https://www.w3.org/TR/rdf-schema/#ch_resource

**Comment by philblackwood** [comment]
marksem, yes, it looks like proxies are a good choice. Attached is a slide relating it to basic set theory.

[ordering literals.pptx](https://github.com/semanticarts/gist/files/6502018/ordering.literals.pptx)

**Comment by rjyounes** [comment]
@marksem Regardless of the fact that `rdfs:Literal`s are also `rdfs:Resource`s, your solution requires `proxyFor` to be an `rdf:Property` rather than an OWL property, which must be either `owl:DatatypeProperty` or `owl:ObjectProperty`.  An object property can only point at an IRI or a bnode, a datatype property only at a literal.

I assume that if we define an `rdf:Property` in gist, gist is no longer an OWL ontology, so we don't want to go that route.

A literal can't have a sequence number assigned to it in any case, so you'd have to either use the `rdf:List` model or "autobox" the literal.

**Comment by marksem** [comment]
Ah yes, a very good point.  My head was in graph land, but not OWL land. :D

**Comment by uscholdm** [comment]
> A literal can't have a sequence number assigned to it in any case, i.e., aside from our proxy model, so you'd have to either use the rdf:List model or "autobox" the literal.

No, but we can always package up any literal as a gist:Text object and use the usual proxy pattern.

**Comment by sa-bpelakh** [comment]
Have we considered a DatatypeProperty equivalent of `isProxyFor`, and each proxy could be defined as having _either_ a proxied object or a proxied literal. I am sure there is clever OWL definition (I know SHACL has a `xone` construct).

**Comment by rjyounes** [comment]
We need to address this issue before a PR is created. @daliadahleh  I am moving it back to in review.

**Comment by rjyounes** [comment]
See comment above where I proposed two `isProxyFor` predicates: `isProxyForLiteral` and `isProxyForObject`. Not completely appealing, but is there another option?

**Comment by marksem** [comment]
The above seems necessary.  The names seem good to me.

**Comment by uscholdm** [comment]
> See comment above where I proposed two isProxyFor predicates: isProxyForLiteral and isProxyForObject. Not completely appealing, but is there another option?

Yes there is. Use gist:Text, assuming we can live with string literals. Like this:
![image](https://user-images.githubusercontent.com/10777545/121127739-15a98e00-c7df-11eb-9fb2-2b3914041d60.png)

**Comment by rjyounes** [comment]
@uschold `gist:Text` has a restriction on `gist:containedText` that it be `xsd:string`. So this won't help with dates, numeric values, etc.

**Comment by rjyounes** [comment]
Also, it's a question of one additional predicate vs. an additional node (and level of indirection) in an instance graph for each collection member.  I think the former is preferable. This nicely illustrates two competing aspects of complexity: (1) number of terms in the ontology, and (2) number of required nodes in an instance graph.

**Comment by uscholdm** [comment]
> Also, it's a question of one additional predicate vs. an additional node (and level of indirection) in an instance graph for each collection member. I think the former is preferable.

Its also a question of how much extra stuff do we put in gist to cater for all conceivable scenarios. We normally cater for the most likely to arise frequently and let the users add extras as they see fit. So we could keep it simple and just have the object property and not the literal one. In a scopeNote indicate how to cater for literals. 

Any

**Comment by rjyounes** [comment]
How do you handle non-string literals, though? This was my point above.

**Comment by uscholdm** [comment]
If you want the number 1, then make it a string `"1"^^xsd:string`. 
You can make dates and any other kind of literal strings tool.

Is there a compelling need for a more robust solution?

**Comment by rjyounes** [comment]
Yes. Dates and numbers should not be converted to strings since then you can't do anything with them. As an example, you can't determine whether the sequencing of dates in the ordered lists corresponds to the actual temporal sequencing without casting.

The main thing is to determine whether there are use cases for dealing with ordered lists of literals in this way. But if there is, I think the additional predicate is a much simpler and more straightforward solution.

**Comment by rjyounes** [comment]
Boris: We tend to wrap literals in objects anyway, so not needed.

**Comment by philblackwood** [comment]
Representing numbers as strings as proposed introduces 3 layers of indirection, which is effectively 3 "proxies".

Start with the number 1, then

1. represent it as a string
2. represent it again as a word
3. represent it again as a proxy

It will be easier to learn and use if there is just one proxy for the number. Introducing a new data property in parallel with the object property isProxyFor will be more like what most modelers would be likely to come up with on their own. 

As a user, I would not mind if this proxy for literals were only included in the model as a scopeNote of the proxy-for object property.
